core/ir/instr_shared.c

```diff
@@ -449,6 +449,20 @@ instr_get_opcode(instr_t *instr)
 /* in rest of file, directly de-reference for performance (PR 622253) */
 #define instr_get_opcode inlined_instr_get_opcode
 
+#define inlined_instr_get_category(instr)                                         \
+    (IF_DEBUG_(CLIENT_ASSERT(sizeof(*instr) == sizeof(instr_t), "invalid type"))( \
+        ((instr)->category == DR_INSTR_CATEGORY_UNCATEGORIZED ||                  \
+         !instr_operands_valid(instr))                                            \
+            ? (instr_decode_with_current_dcontext(instr), (instr)->category)      \
```

Comment:
But as mentioned in the prior review round, what about a synthetic instr with no raw bits?  Does an encoding have to be done (cached in the raw bits) and then a decoding?  At the very least, document that this API is not supported for synthetic instructions and is only supported for decoded instructions (and someone can add synthetic support later: document the lack of support in the header, and here put a `XXX i#6238: ...` comment about it).